src/typing/flow_js.ml

```diff
@@ -4870,14 +4870,16 @@ let rec __flow cx ((l: Type.t), (u: Type.use_t)) trace =
           |> replace_reason (fun desc -> RPropertyOf (x, desc))
           |> repos_reason (aloc_of_reason reason_op)
         in
-        match Property.read_t p with
-        | Some t ->
-          let propref = Named (reason_prop, x) in
-          let t = filter_optional cx ~trace reason_prop t in
-          rec_flow cx trace (to_obj, SetPropT (
-            unknown_use, reason_prop, propref, Normal, t, None
-          ))
-        | None ->
+        match Property.read_t p, to_obj with
+        | Some _, (DefT (_, ObjT ob)) when not ob.flags.frozen ->
+          Context.set_prop cx ob.props_tmap x p
+        | Some t, to_obj ->
```

Comment:
this branch was still needed for these to work: https://github.com/facebook/flow/blob/master/tests/destructuring/destructuring.js#L61
https://github.com/facebook/flow/blob/master/tests/object_assign/non_objects.js

Reply:
https://github.com/facebook/flow/blob/master/tests/destructuring/destructuring.js#L61 does not work without `| Some t, to_obj ->` branch because [this](https://github.com/facebook/flow/blob/99e058a7d3f2fc6f816ee9d8616e0814a91932cb/src/typing/flow_js.ml#L4987) generates OpenT instead of ObjT. Otherwise `| Some t, to_obj ->` branch could be unnecessary altogether.